source/vision/visionHandler.py

```diff
@@ -0,0 +1,194 @@
+#vision/visionHandler.py
+#A part of NonVisual Desktop Access (NVDA)
+#This file is covered by the GNU General Public License.
+#See the file COPYING for more details.
+#Copyright (C) 2018-2019 NV Access Limited, Babbage B.V.
+
+"""Module containing the vision handler.
+"""
+
+from .constants import *
+from .providerBase import VisionEnhancementProvider
+from .visionHandlerExtensionPoints import EventExtensionPoints
+import pkgutil
+import importlib
+from baseObject import AutoPropertyObject
+import api
+import config
+import weakref
+from logHandler import log
+import wx
+from collections import defaultdict
+import copy
+
+def getProviderClass(moduleName, caseSensitive=True):
+	"""Returns a registered provider class with the specified moduleName."""
+	try:
+		return __import__(
+			"visionEnhancementProviders.%s" % moduleName,
+			globals(),
+			locals(),
+			("visionEnhancementProviders",)
+		).VisionEnhancementProvider
+	except ImportError as initialException:
+		if caseSensitive:
+			raise initialException
+		for loader, name, isPkg in pkgutil.iter_modules(visionEnhancementProviders.__path__):
+			if name.startswith('_') or name.lower() != moduleName.lower():
+				continue
+			return __import__(
+				"visionEnhancementProviders.%s" % name,
+				globals(),
+				locals(),
+				("visionEnhancementProviders",)
+			).VisionEnhancementProvider
+		else:
+			raise initialException
+
+class VisionHandler(AutoPropertyObject):
+
+	def __init__(self):
+		self.lastReviewMoveContext = None
+		self.lastCaretObjRef = None
+		self.providers = dict()
+		self.extensionPoints = EventExtensionPoints()
+		# Handle first initialization of the handler as a config profile switch.
+		# However, execute this on the main thread, which makes sure that
+		# the gui is fully initialized before providers are initialized that might rely on it.
+		wx.CallAfter(self.handleConfigProfileSwitch)
+		config.post_configProfileSwitch.register(self.handleConfigProfileSwitch)
+
+	def terminateProvider(self, providerName):
+		providerInstance =self.providers.pop(providerName, None)
+		if not providerInstance:
+			log.warning("Tried to terminate uninitialized provider %s" % providerName)
+			return False
+		try:
+			providerInstance.terminate()
+		except:
+			# Purposely catch everything.
+			# A provider can raise whatever exception,
+			# therefore it is unknown what to expect.
+			log.error("Error while terminating vision provider %s" % providerName, exc_info=True)
+			return False

```

Comment:
Beware, this leaves the `providers` list missing `providerInstance`, but returns indicating that it can't be removed. The other `return False` has the same issue. This may be intentional, in which case I think it deserves a comment about why it's what we want.